src/state/signalr-connection.ts

```diff
@@ -0,0 +1,139 @@
+import _Vue from 'vue';
+import { LogLevel, HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
+import { StateManager } from '@/state';
+import { SignalRUtilities, NotificationHandler } from '@/dtos/temp-signalr';
+
+type ConnectionStatus = 'closed' | 'connecting' | 'connected' | 'closing';
+
+export class SignalRWrapper {
+    // Manage the SignalR connection. This is a singleton class
+    private static _instance: SignalRWrapper;
+
+    private _connection?: HubConnection;
+    private _utils?: SignalRUtilities;
+    private _currentHandler?: NotificationHandler;
+    private _status: ConnectionStatus = 'closed';
+    private _subscribedEditionId?: number;
+
+    public static get instance() {
+        if (!SignalRWrapper._instance) {
+            SignalRWrapper._instance = new SignalRWrapper();
+        }
+
+        return SignalRWrapper._instance;
+    }
+
+    private constructor() {
+        if (!process.env.VUE_APP_SIGNALR_URL) {
+            throw new Error('VUE_APP_SIGNALR_URL not defined, please add it to the environment');
+        }
+    }
+
+    public async subscribeEdition(editionId: number) {
+        console.debug('SignalR subscribeEdition called, current status ', this._status);
+        if (this._subscribedEditionId === editionId) {
+            return;
+        }
+
+        if (this._status !== 'connected') {
+            await this.connect();
+        }
+
+        if (this._subscribedEditionId) {
+            await this.unsubscribeEdition();
+        }
+
+        await this._utils!.subscribeToEdition(editionId);
+        console.debug('SubscribeToEdition called');
+        this._subscribedEditionId = editionId;
+    }
+
+    public async unsubscribeEdition() {
+        console.debug('SignalR unsubscribeEdition, status is ', this._status);
+        if (this._status !== 'connected' || !this._subscribedEditionId) {
+            return;
+        }
+
+        await this._utils!.unsubscribeToEdition(this._subscribedEditionId);
+        console.debug('UnsubscribeToEdition called');
+        this._subscribedEditionId = undefined;
+    }
+
+    public async userChanged() {
+        console.debug('SignalR changing users');
+        const subscribed = this._subscribedEditionId;
+        await this.disconnect();
+        await this.connect();
+        if (subscribed) {
+            await this.subscribeEdition(subscribed);
+        }
+    }
+
+    public registerNotificationHandler(handler: NotificationHandler) {
+        this.unregisterNotificationHandler();
+        this._currentHandler = handler;
+        if (this._utils) {
+            this._utils!.connectNotificationHandler(handler);
+        }
+    }
+
+    public unregisterNotificationHandler() {
+        if (this._currentHandler && this._utils) {
+            this._utils.disconnectNotificationHandler(this._currentHandler);
+        }
```

Comment:
I was a little confused about two things here.  Firstly the usage of this._currentHandler implies that there will only be one handler at a time running. Certainly we will need more than one sometimes.  Secondly, I think with the this._currentHandler variable you were rightly trying to make sure that you are calling "off" with a reference to the same function that "on" was called with (which you must do).  So I guess we probably need here some sort of array or object to hold all the currently registered NotificationHandlers.  You probably need this so that you can reregister the listeners on reconnect (check the docs for the new policy on autoreconnect for the v3.x signalr javascript client: https://docs.microsoft.com/en-us/aspnet/core/signalr/javascript-client?view=aspnetcore-3.1#reconnect-clients).